### PR TITLE
[FSDP] Relax `sharded_grad` assert to allow IDLE

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96584

`_use_sharded_grad_views()` can be called when re-registering the original parameters in `load_state_dict()`, in which case the training state is `IDLE`. Previously, I only expected `_use_sharded_grad_views()` to be called in `FORWARD` when the sharded gradient is not in `_saved_grad_shard` or `_cpu_grad`.

### CODE REVIEW

**Review by fegin — APPROVED** [review]
LGTM

**Review by zhaojuanmao — APPROVED** [review]

**Comment by awgu** [comment]
Failure not related:
```
2023-03-13T19:06:23.2287711Z ERROR: Could not find a version that satisfies the requirement pytorch-triton==2.1.0+b8b470bc59 (from torch) (from versions: 0.0.1, 2.0.0+0d7e753227, 2.0.0+3aa3d7024e, 2.0.0+af76c989eb, 2.0.0+b8b470bc59, 2.0.0+c8bfe3f548, 2.0.0+d54c04abe2, 2.1.0+2c32f43999)
2023-03-13T19:06:23.2290161Z ERROR: No matching distribution found for pytorch-triton==2.1.0+b8b470bc59
```
